core/services/ping/portwatcher.py

```diff
@@ -0,0 +1,91 @@
+import asyncio
+import logging
+from typing import Callable, Dict, List, Optional, Set
+from warnings import warn
+
+import serial.tools.list_ports
+from serial.tools.list_ports_linux import SysFS
+
+from pingutils import PingDeviceDescriptor
+
+MAX_ATTEMPTS = 10
+
+
+def pingDevicesFilter(port: SysFS) -> bool:
+    """Accepts ports with known chips in the Ping1D and Ping360. Internal ports
+    Are handled by Ardupilot and show None for interface, so we ignore those too.
+    """
+    if not port.interface:
+        return False
+    return "FT231X" in port.interface or "FT230" in port.interface
+
+
+class PortWatcher:
+    """Watches the Serial ports on the system, calls set_prober when a port is found,
+    port_post_callback when a port is no longe present"""
+
+    def __init__(self) -> None:
+        logging.info("PortWatcher Started")
+        self.known_ports: Set[str] = set()
+        self.probe_callback: Optional[Callable[[SysFS], Optional[PingDeviceDescriptor]]] = None
+        self.port_lost_callback: Optional[Callable[[SysFS], None]] = None
+        self.filters: List[Callable[[SysFS], bool]] = [pingDevicesFilter]
+        self.probe_attempts_counter: Dict[SysFS, int] = {}
+
+    def set_prober(self, callback: Callable[[SysFS], Optional[PingDeviceDescriptor]]) -> None:
+        self.probe_callback = callback
+
+    def set_port_post_callback(self, callback: Callable[[SysFS], None]) -> None:
+        self.port_lost_callback = callback
+
+    def port_should_be_probed(self, port: SysFS) -> bool:
+        """A port should be probed if there hasn't been MAX_ATTEMPTS to probe it yet
+        and it is caught by our filters
+        """
+        if self.probe_attempts_counter.get(port, 0) >= MAX_ATTEMPTS:
+            return False
+        return any(filter(port) for filter in self.filters)
```

Comment:
this is a bit weird since filter is also an python built-in function. maybe `port_filter(port)` ?

Reply:
can we get a linter to catch these?
somethign like this?
https://pypi.org/project/flake8-builtins/

Reply:
We are already testing such thing, are you running or pre-push file ?
You should have: 
```
W0622: Redefining built-in 'filter' (redefined-builtin)
```

Reply:
yes, and it also passed CI...

Reply:
I have created an issue, it appears to be a bug.
https://github.com/PyCQA/pylint/issues/4472

Reply:
I'm wondering if maybe it doesn't count because its scope is just that line

Reply:
Ok, but we should just update the name to avoid such problem when possible, otherwise is a bit confuse.

Reply:
sure, done.